sia/host/contracts.go

```diff
@@ -0,0 +1,214 @@
+	if fileSize < h.announcement.MinFilesize || fileSize > h.announcement.MaxFilesize {
+		err = fmt.Errorf("file is of incorrect size - filesize %v, min %v, max %v", fileSize, h.announcement.MinFilesize, h.announcement.MaxFilesize)
+		return
+	}
+	// Check that there is space for the file.
+	if fileSize > uint64(h.spaceRemaining) {
+		err = errors.New("host is at capacity and can not take more files.")
+		return
+	}
+	// Check that the duration of the contract is in bounds.
+	if fullDuration < h.announcement.MinDuration || fullDuration > h.announcement.MaxDuration {
+		err = errors.New("contract duration is out of bounds")
+		return
+	}
+	// Check that challenges will not be happening too frequently or infrequently.
+	if t.FileContracts[0].ChallengeWindow < h.announcement.MinChallengeWindow || t.FileContracts[0].ChallengeWindow > h.announcement.MaxChallengeWindow {
+		err = errors.New("challenges frequency is too often")
+		return
+	}
+	// Check that tolerance is acceptible.
+	if t.FileContracts[0].Tolerance < h.announcement.MinTolerance {
+		err = errors.New("tolerance is too low")
+		return
+	}
+	// Outputs for successful proofs need to go to the correct address.
+	if t.FileContracts[0].ValidProofAddress != h.announcement.CoinAddress {
+		err = errors.New("coins are not paying out to correct address")
+		return
+	}
+	// Outputs for successful proofs need to match the price.
+	requiredSize := h.announcement.Price * consensus.Currency(fileSize) * consensus.Currency(t.FileContracts[0].ChallengeWindow)
+	if t.FileContracts[0].ValidProofPayout < requiredSize {
+		err = errors.New("valid proof payout is too low")
+		return
+	}
+	// Output for failed proofs needs to be the 0 address.
+	emptyAddress := consensus.CoinAddress{}
+	if t.FileContracts[0].MissedProofAddress != emptyAddress {
+		err = errors.New("burn payout needs to go to the empty address")
+		return
+	}
+	// Verify that output for failed proofs matches burn.
+	maxBurn := h.announcement.Burn * consensus.Currency(fileSize) * consensus.Currency(t.FileContracts[0].ChallengeWindow)
+	if t.FileContracts[0].MissedProofPayout > maxBurn {
+		err = errors.New("burn payout is too high for a missed proof.")
+		return
+	}
+	// Verify that the contract fund covers the payout and burn for the whole
+	// duration.
+	requiredFund := (h.announcement.Burn + h.announcement.Price) * consensus.Currency(fileSize) * consensus.Currency(contractDuration)
+	if t.FileContracts[0].ContractFund < requiredFund {
+		err = errors.New("ContractFund does not cover the entire duration of the contract.")
+		return
+	}
+
+	// Add enough funds to the transaction to cover the penalty half of the
+	// agreement.
+	penalty := h.announcement.Burn * consensus.Currency(fileSize) * consensus.Currency(contractDuration)
+	id, err := h.wallet.RegisterTransaction(t)
+	if err != nil {
+		return
+	}
+	err = h.wallet.FundTransaction(id, penalty)
+	if err != nil {
+		// TODO: This leaks that the host is out of money.
+		return
+	}
```

Comment:
hmm. This is tricky. I assume there are good reasons why we wouldn't want to leak that info. We could return a generic error, like `"contract refused"` or `"internal error"`, but even those would be a giveaway for anyone who's read the source code.